sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/SingleQueueingConsumer.java

```diff
@@ -0,0 +1,156 @@
+ *   <li>Channel thread will block for up to a configurable length of time (default: 10 minutes) for
+ *       the queue to have an open slot. Previous it would nominally fail immediately, which would
+ *       never happen as the queue was unbounded.
+ *   <li>Exceptions have been simplified to only expose IOException to callers.
+ * </ul>
+ *
+ * <p>Note: setting amqp prefetch to 1 does not accomplish the same thing. Prefetch limits the
+ * number of messages that will be sent to the consumer (good) *which have not been acknowledged
+ * yet* (bad for Beam). Because acknowledgements happen during {@code finalizeCheckpoint}, prefetch
+ * would have to be set to the maximum number of messages Beam *could* process before {@code
+ * finalizeCheckpitn} is called, which is unknowable.
+ *
+ * <p>In effect, this version will use less memory than the original but result in lower throughput
+ * among any single beam shard.
+ *
+ * @see
+ *     "https://github.com/rabbitmq/rabbitmq-java-client/blob/4.x.x-stable/src/main/java/com/rabbitmq/client/QueueingConsumer.java"
+ *     for the original QueuingConsumer
+ */
+public class SingleQueueingConsumer extends DefaultConsumer {
+  private final BlockingDeque<Delivery> queue = new LinkedBlockingDeque<>(1);
+  private final long offerTimeoutMillis;
+
+  private static final int DEFAULT_DELIVER_TIMEOUT_MINUTES = 10;
+
+  private volatile @Nullable ShutdownSignalException shutdown = null;
+  private volatile boolean cancelled = false;
+
+  /**
+   * Default implementation which will block delivering messages for up to 10 minutes if the queue
+   * is full.
+   */
+  public SingleQueueingConsumer(Channel channel) {
+    this(channel, DEFAULT_DELIVER_TIMEOUT_MINUTES, TimeUnit.MINUTES);
+  }
+
+  /**
+   * Blocks deliveries for the supplied length of time when the queue is full.
+   *
+   * @param channel
+   * @param deliverTimeout
+   * @param deliverTimeoutUnits
+   */
+  public SingleQueueingConsumer(
+      Channel channel, long deliverTimeout, TimeUnit deliverTimeoutUnits) {
+    super(channel);
+    this.offerTimeoutMillis = deliverTimeoutUnits.toMillis(deliverTimeout);
+  }
+
+  @Override
+  public void handleCancelOk(String consumerTag) {
+    doCancel();
+  }
+
+  @Override
+  public void handleCancel(String consumerTag) throws IOException {
+    doCancel();
+  }
+
+  private void doCancel() {
+    this.cancelled = true;
+  }
+
+  @Override
+  public void handleShutdownSignal(String consumerTag, ShutdownSignalException sig) {
+    this.shutdown = sig;
+  }
+
+  @Override
+  public void handleDelivery(
+      String consumerTag, Envelope envelope, AMQP.BasicProperties properties, byte[] body)
+      throws IOException {
+    if (shutdown != null) {
+      throw new IOException(shutdown);
+    }
+
+    Delivery delivery = new Delivery(envelope, properties, body);
+    try {
+      queue.offer(delivery, offerTimeoutMillis, TimeUnit.MILLISECONDS);
```

Comment:
This is likely to be the only contentious part of this PR: 1) how long to block the delivery of a message (or: the maximum time it should take the pipeline to process a single message); and 2) what to do if it takes longer than this.

The current implementation defaults to waiting 10 minutes for a message to be processed (`poll` to be called here). And if it fails to do so, the channel will be closed, effectively stopping the pipeline.